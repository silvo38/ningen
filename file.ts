import { path } from "./deps.ts";

export type FileFn = (path: string) => File;

/**
 * Returns a list of files. Supply either a single `string[]`, or a variable
 * number of strings.
 */
export type FilesFn = (...filenames: (string | string[])[]) => Files;

export class File {
  constructor(private readonly path: string) {}

  getAbsolutePath(): string {
    return this.path;
  }

  getRelativePath(directory: string): string {
    return path.relative(directory, this.path);
  }

  withSuffix(suffix: string): File {
    return new File(this.path + suffix);
  }
}

export type Files = readonly File[];

/**
 * Constructs a {@link File} object from the given filename. If {@code filename}
 * is a relative path, uses {@code directory} as the root directory.
 */
export function file(directory: string, filename: string): File {
  if (path.isAbsolute(filename)) {
    return new File(filename);
  } else {
    return new File(path.join(directory, filename));
  }
}

/**
 * Constructs a list of {@link File} objects from the given filenames. You can
 * supply either an array of filenames, or a vararg list of filenames.
 */
export function files(
  directory: string,
  ...filenames: (string | string[])[]
): Files {
  if (filenames.length == 1) {
    const elem = filenames[0];
    if (typeof elem === "string") {
      // Single string.
      return [file(directory, elem)];
    } else {
      // Single array of strings.
      return elem.map((f) => file(directory, f));
    }
  }
  // Vararg list of (hopefully) strings. Throw if not.
  return filenames.map((f, index) => {
    if (typeof f !== "string") {
      throw new Error(
        `Expected list of strings but element at index ${index} was ${f}`,
      );
    }
    return file(directory, f);
  });
}
