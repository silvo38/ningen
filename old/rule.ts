import { checkRepoRelativePath, sortedUnique } from "./util.ts";

export interface RuleOpts {
  /** The name for the rule in the generated ninja file. */
  name: string;

  /** The command to run. */
  command: string;

  /**
   * Specifies that the rule is used to generate the ninja build file. Defaults
   * to false.
   */
  generator?: boolean;

  /**
   * Optional implicit inputs that will be added to every build target that uses
   * the rule, e.g. you might put the executable used in the command here, or a
   * shell script for it, etc.
   */
  implicit?: string[];

  /** Optional path for a generated gcc-style depfile produced by the rule. */
  depfile?: string;
}

export class Rule {
  readonly name: string;
  readonly command: string;
  readonly implicit: ReadonlyArray<string>;
  readonly generator: boolean;
  readonly depfile?: string;

  constructor({ name, command, implicit, generator, depfile }: RuleOpts) {
    this.name = name;
    this.command = command;
    this.generator = generator ?? false;
    this.implicit = sortedUnique(implicit ?? []);
    this.depfile = depfile;

    this.implicit.forEach((f) => checkRepoRelativePath(f));
  }
}
