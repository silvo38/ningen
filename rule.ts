import { File, Files } from "./file.ts";

export type RuleFn = (
  opts: {
    name: string;
    command: string;
    binary?: File;
    srcs?: Files;
    vars?: Vars;
  },
) => Rule;

/** Variables usable in the shell command for a rule. */
export type Vars = Record<string, string | File>;

export class Rule {
  constructor(
    readonly name: string,
    readonly command: string,
    readonly binary: File | null,
    readonly srcs: Files,
    readonly vars: Vars,
  ) {}
}

export function rule(
  { name, command, binary, srcs, vars }: {
    /** The name for the rule. */
    name: string;
    /** The shell command to invoke. */
    command: string;
    /**
     * An optional binary name or path to invoke. Available in the shell command
     * as a special `$binary` variable.
     */
    binary?: File;
    /** Optional implicit deps to add to build targets using this rule. */
    srcs?: Files;
    /**
     * Optional environment variables set to their default values. These can be
     * used in the command, and can be overridden in individual build targets.
     */
    vars?: Vars;
  },
): Rule {
  srcs = srcs ?? [];
  vars = vars ?? {};
  if (binary != null) {
    // Add bin to srcs if missing, if it's a File.
    if (typeof binary != "string" && !srcs.some((f) => f.equals(binary))) {
      srcs = [...srcs, binary];
    }
  }
  return new Rule(name, command, binary ?? null, srcs, vars);
}
