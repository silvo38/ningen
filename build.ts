import { Files } from "./file.ts";
import { Rule } from "./rule.ts";

export type BuildFn = (
  { rule, inputs, outputs }: {
    rule: Rule;
    inputs: Files;
    outputs: Files;
  },
) => void;

export class Target {
  constructor(
    readonly rule: Rule,
    readonly inputs: Files,
    readonly outputs: Files,
    readonly implicit: Files,
  ) {}
}

export function build(
  { rule, inputs, outputs }: {
    rule: Rule;
    inputs: Files;
    outputs: Files;
  },
): Target {
  const implicit = rule.srcs;
  return new Target(rule, inputs, outputs, implicit);
}
