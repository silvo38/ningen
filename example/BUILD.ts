#!/usr/bin/env -S deno run --allow-read --allow-write --unstable
// The shebang above lets you execute this BUILD.ts file directly.

// TODO: Give real deno.land URL.
import { File, init, Target } from "../mod.ts";

// Initialises Ningen and retrieves the functions for defining rules and build
// targets.
const { build, rule, file, files, generate } = init(import.meta.url);

// Define a Rule for the `append.sh` script.
const appendRule = rule({
  name: "append",
  command: "./append.sh $in $out",
  srcs: files("append.sh"),
});

// Define a helper function that invokes the rule.
function append(src: File) {
  build({
    rule: appendRule,
    inputs: [src],
    outputs: [src.withSuffix(".out")],
  });
}

// Process a file.
append(file("abc.txt"));

// Write the build.ninja file. Can override the output file if you like.
generate();
