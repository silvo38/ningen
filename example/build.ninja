rule append
  command = ./append.sh $in $out
rule ningen
  command = ./BUILD.ts
  generator = 1

build abc.txt.out: append abc.txt | append.sh
build build.ninja: ningen BUILD.ts
