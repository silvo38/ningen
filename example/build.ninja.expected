rule append
  command = ./append.sh $in $out

build abc.txt.out: append abc.txt | append.sh
